x/auth/ante.go

```diff
@@ -194,18 +202,24 @@ func processPubKey(acc Account, sig StdSignature, simulate bool) (crypto.PubKey,
 	return pubKey, sdk.Result{}
 }
 
-func consumeSignatureVerificationGas(meter sdk.GasMeter, pubkey crypto.PubKey) {
-	switch pubkey.(type) {
-	case ed25519.PubKeyEd25519:
-		meter.ConsumeGas(ed25519VerifyCost, "ante verify: ed25519")
-	case secp256k1.PubKeySecp256k1:
-		meter.ConsumeGas(secp256k1VerifyCost, "ante verify: secp256k1")
+// consumeSignatureVerificationGas consumes gas for signature verification based
+// upon the public key type. The cost is fetched from the given params and is
+// matched by the concrete type.
+//
+// TODO: Design a cleaner and flexible way to match concrete public key types.
+func consumeSignatureVerificationGas(meter sdk.GasMeter, pubkey crypto.PubKey, params Params) {
+	pubkeyType := strings.ToLower(fmt.Sprintf("%T", pubkey))
+	switch {
+	case strings.Contains(pubkeyType, "ed25519"):
+		meter.ConsumeGas(params.SigVerifyCostED25519, "ante verify: ed25519")
+	case strings.Contains(pubkeyType, "secp256k1"):
```

Comment:
This should not be like this. I may want ristretto-ed25519 which should have different gas than normal ed25519. Instead it should be dynamically getting the pubkey type (perhaps amino route) and querying param store with that. Basically blocks on: https://github.com/tendermint/go-amino/issues/231

I don't feel that comfortable with this as is. (Especially since it makes the state machine dependent on the tendermint struct name)

Reply:
Ok, but what if I don't want to use Amino? And how do I query a param store with a dynamic type/field/key (e.g. how do I ask, "get me the cost of key type X"?). I don't think we can use maps afaik.

Reply:
We could add a name or type function to the pubkey interface then.

Reply:
Yeah! I think that is a neat idea. Question still looms about how to get such dynamic values. 

Not the ideal, but slightly better interim approach could be to have all key types (e.g. ristretto-ed25519, ed25519, secp256k1, etc...) that we desire to be defined as first class fields in the params. Not flexible, but better than what we have. Thoughts @ValarDragon?

Reply:
@ValarDragon in auth (or somewhere else), we can define:

```go
type TypedPubKey interface {
	tmcrypto.PubKey

	Type() string
}
```

As for the params, I'm still having trouble finding where your solution is stated?

Reply:
:+1: Typed PubKey! For the param store, the key would then be TypedPubKey.Type()

Reply:
Actually I do think named pubkey is a better name, since this about names for the pubkey type, rather than some sort of type system thing.

Reply:
@ValarDragon do you mind if we keep this for now? I have no clue when something like would get into TM? Also Ethermint is dependent on this PR, which means now it would be dependent on two PRs in two different repos.

Reply:
This does add quite a bit of technical debt, although the struct names are unlikely to change. At minimum let's create an issue upstream.

Reply:
Created an issue here: https://github.com/tendermint/tendermint/issues/3050